### PR TITLE
Dev branch merge before dmenu integration

### PR DESCRIPTION
**Pull request by Toasterbirb**

### CODE REVIEW

File: README [hunk 1]
```diff
@@ -1,15 +1,20 @@
-dwm - dynamic window manager
-============================
-dwm is an extremely fast, small, and dynamic window manager for X.
-
-
-Requirements
-------------
+# dde - dynamic desktop environment
```
Comment (thatnerdjosh) [inline README:1]: I think we should add a file extension on this, it is clearly using some sort of formatting (looks like markdown).

if markdown, maybe `.md` would be appropriate. This would allow us to later provide linting in CI, and also helps for editor integration (tools such as `.editorconfig` and `markdownlint`)

File: config.def.h [hunk 2]
```diff
@@ -1,6 +1,13 @@
 /* See LICENSE file for copyright and license details. */
 
 /* appearance options */
+
+/* wallpaper */
+static const bool restorewallpaper 	= false; 	/* when enabled, the wallpapercmd command is run */
```
Comment (thatnerdjosh) [inline config.def.h:6]: This seems like a hack and doesn't feel like it belongs here. Wouldn't be be better to provide a sample `.xinitrc` (or some other scripts) to achieve this same thing?

File: config.def.h [hunk 3]
```diff
@@ -74,53 +81,61 @@ static const Layout layouts[] = {
...
 	{ MODKEY,                       XK_k,      focusstack,     {.i = -1 } },
-	{ MODKEY,                       XK_i,      incnmaster,     {.i = +1 } },
-	{ MODKEY,                       XK_d,      incnmaster,     {.i = -1 } },
+	{ MODKEY|ControlMask,           XK_i,      incnmaster,     {.i = +1 } },
+	{ MODKEY|ControlMask,           XK_d,      incnmaster,     {.i = -1 } },
 	{ MODKEY,                       XK_h,      setmfact,       {.f = -0.05} },
 	{ MODKEY,                       XK_l,      setmfact,       {.f = +0.05} },
-	{ MODKEY,                       XK_Return, zoom,           {0} },
+	{ MODKEY|ShiftMask,             XK_Return, zoom,           {0} },
+	{ MODKEY, 			            XK_f,      togglefullscr,  {0} },
```
Comment (thatnerdjosh) [inline config.def.h:100]: spacing

File: config.def.h [hunk 4]
```diff
@@ -74,53 +81,61 @@ static const Layout layouts[] = {
...
 	{ MODKEY,                       XK_Tab,    view,           {0} },
-	{ MODKEY|ShiftMask,             XK_c,      killclient,     {0} },
+	{ MODKEY|ShiftMask,             XK_q,      killclient,     {0} },
+
+	/* layouts */
+	{ MODKEY, 						XK_y, 	   layoutmenu, 	   {0} },
 	{ MODKEY,                       XK_t,      setlayout,      {.v = &layouts[0]} },
-	{ MODKEY,                       XK_f,      setlayout,      {.v = &layouts[1]} },
+	//{ MODKEY,                       XK_f,      setlayout,      {.v = &layouts[1]} },
```
Comment (thatnerdjosh) [inline config.def.h:107]: why?

File: config.def.h [hunk 5]
```diff
@@ -74,53 +81,61 @@ static const Layout layouts[] = {
...
-	{ MODKEY,                       XK_o,      incrihgaps,     {.i = -1 } },
-	{ MODKEY|ControlMask,           XK_y,      incrivgaps,     {.i = +1 } },
-	{ MODKEY|ControlMask,           XK_o,      incrivgaps,     {.i = -1 } },
-	{ MODKEY|Mod4Mask,              XK_y,      incrohgaps,     {.i = +1 } },
-	{ MODKEY|Mod4Mask,              XK_o,      incrohgaps,     {.i = -1 } },
-	{ MODKEY|ShiftMask,             XK_y,      incrovgaps,     {.i = +1 } },
-	{ MODKEY|ShiftMask,             XK_o,      incrovgaps,     {.i = -1 } },
+	{ MODKEY|Mod1Mask,              XK_h,      incrgaps,       {.i = +4 } },
+	{ MODKEY|Mod1Mask,              XK_l,      incrgaps,       {.i = -4 } },
+	//{ MODKEY|Mod1Mask|ShiftMask,    XK_h,      incrogaps,      {.i = +4 } },
```
Comment (thatnerdjosh) [inline config.def.h:127]: I don't see a reason to leave these in the `config.def.h`

File: dwm.cpp [hunk 6]
```diff
@@ -911,6 +915,7 @@ manage(Window w, XWindowAttributes *wa)
 	wc.border_width = c->bw;
 	XConfigureWindow(dpy, w, CWBorderWidth, &wc);
 	XSetWindowBorder(dpy, w, scheme[SchemeNorm][ColBorder].pixel);
+	XSetWindowBorder(dpy, w + 50, scheme[SchemeSel][ColBorder].pixel);
```
Comment (thatnerdjosh) [inline dwm.cpp:918]: what does it do?

File: dwm.cpp [hunk 7]
```diff
@@ -1473,6 +1478,47 @@ setlayout(const Arg *arg)
 		drawbar(selmon);
 }
 
+void layoutmenu(const Arg *arg)
+{
+	/* How many symbols can one layout name fit including the number */
+	int maxlayoutsymbolsize = 10;
+
+	int layoutcount = (sizeof(layouts)/sizeof(*layouts));
```
Comment (thatnerdjosh) [inline dwm.cpp:1486]: how does this work?

File: dwm.cpp [hunk 8]
```diff
@@ -1473,6 +1478,47 @@ setlayout(const Arg *arg)
...
+	char layoutlist[layoutcount * maxlayoutsymbolsize];
+	sprintf(layoutlist, "1. %s", layouts[0].symbol);
```
Comment (thatnerdjosh) [inline dwm.cpp:1488]: why is this needed vs just generating the entire string via the loop

**Comment by thatnerdjosh** [comment]
LGTM aside from the comments and one note:

it may be worth extracting out that layout menu to a separate file so it is easy to identify what to extract out should we choose to later go down the session manager route (such as LXSession)

**Review by thatnerdjosh — APPROVED** [review]

File: config.def.h (hunk 4 above)
Comment (Toasterbirb) [reply config.def.h:107]: quite possibly a temporary solution, it has been uncommented in more recent version

File: dwm.cpp (hunk 6 above)
Comment (Toasterbirb) [reply dwm.cpp:918]: literally nothing ;)

File: dwm.cpp (hunk 7 above)
Comment (Toasterbirb) [reply dwm.cpp:1486]: It gets the size of the layouts array. maxlayoutsymbolsize is just a size for buffers used in this function

File: dwm.cpp (hunk 8 above)
Comment (Toasterbirb) [reply dwm.cpp:1488]: Because the lines in the loop need \n newline char in the beginning of the line and the first line is an exception
